can/zynqmp-can/zynqmp-can-test.c

```diff
@@ -0,0 +1,398 @@
+			for (int n = 0; n < buf[i].len; n++) {
+				printf("0x%02X ", buf[i].payload.bytes[n]);
+			}
+			printf("\n");
+		}
+	}
+	else if (ret == -ETIME) {
+		printf("zynqmp-can-test: Received zero frames, timeout\n");
+	}
+	else {
+		printf("zynqmp-can-test: Failed to receive frames, error %i\n", ret);
+	}
+
+	return ret;
+}
+
+/**
+ * Enables or disables canonical mode and echo for the terminal.
+ */
+static void switchmode(int canon)
+{
+	struct termios state;
+
+	tcgetattr(STDIN_FILENO, &state);
+	if (canon) {
+		state.c_lflag |= ICANON;
+		state.c_lflag |= ECHO;
+	}
+	else {
+		state.c_lflag &= ~ICANON;
+		state.c_lflag &= ~ECHO;
+		state.c_cc[VMIN] = 1;
+	}
+	tcsetattr(STDIN_FILENO, TCSANOW, &state);
+}
+
+/**
+ * Receives CAN frames in a loop with a timeout.
+ * Displays received frame statistics in real time.
+ */
+static int recv_in_loop(oid_t *oid)
+{
+	switchmode(0);
+
+	if (mutexCreate(&mutex_stats_access) != 0) {
+		printf("cantest: Failed to create mutex\n");
+		return -1;
+	}
+
+	/* Allocate buffer for received frames */
+	static zynqmp_canFrame buf[64] = { 0 };
+	uint32_t recv_frames = 0;
+
+	/* Start thread to print results each second */
+	handle_t thread_hdl;
+	uint8_t stack[4096];
+	beginthread(cantest_thread, 4, stack, sizeof(stack), &thread_hdl);
+
+	/* Receive frames in a loop with timeout so the loop can be exited */
+	while (run) {
+		time_t timeoutUs = 50 * 1000;
+		int ret = zynqmp_can_recv(oid, buf, ARRAY_LENGTH(buf), timeoutUs, &recv_frames);
+		if (ret == -ETIME) {
+			/* Timeout, ignore and keep trying to read frames (if run is 1) */
+		}
+		else if (ret == EOK) {
+			if (recv_frames == 0) {
+				printf("zynqmp-can-test: Failed, received 0 frames\n");
+			}
+			for (uint32_t i = 0; ((i < ARRAY_LENGTH(buf)) && (i < recv_frames)); i++) {
+				/* Access statistics data displayed by another thread */
+				mutexLock(mutex_stats_access);
+				/* Update statistics for all possible frames */
+				if (buf[i].id > ZYNQMP_CAN_ID_MAX) {
+					printf("zynqmp-can-test: Failed, CAN ID too large\n");
+				}
+				else {
+					frame_stats[buf[i].id].receivedFramesCount++;
+					memcpy(&frame_stats[buf[i].id].lastFrame, &buf[i], sizeof(zynqmp_canFrame));
+				}
+				mutexUnlock(mutex_stats_access);
+			}
+		}
+		else {
+			printf("zynqmp-can-test: Failed to receive frames, error %i\n", ret);
+			return ret;
+		}
+	}
+
+	switchmode(1);
+
+	return 0;
+}
+
+/**
+ * Main function to execute CAN test operations.
+ */
+int main(int argc, char **argv)
+{
+	/* Lookup hardcoded CAN device */
+	char *canDevice = "/dev/can1";
+	oid_t oid;
+	if (lookup(canDevice, NULL, &oid) < 0) {
+		printf("cantest: Cannot lookup device /dev/can1\n");
+		return -1;
+	}
+
+	/* Check if the correct number of arguments was passed */
+	if (argc != 2) {
+		printf("usage: zynqmp-can-test <operation>\n");
+		return -1;
+	}
+
+	/* Perform action based on command */
+	if (0 == strcmp("send", argv[1])) {
```

Comment:
I get the idea behind reversed order in comparisons, but:
a) It's incoherent with our code base,
b) it's incoherent with this file (even the function)
c) gcc emits a warning on `=` instead of `==` for a few years now